storage/s3.go

```diff
@@ -382,6 +386,175 @@ func (s *S3) Get(
+					Err: err,
+				}
+			}
+			return
+		}
+		resultch <- &SelectRecord{
+			Record: record,
+		}
+	}
+	err := resp.EventStream.Err()
+	if err != nil {
+		resultch <- &SelectRecord{
+			Err: err,
+		}
+	}
+}
+
+func (s *S3) selectWorker(ctx context.Context, input s3.SelectObjectContentInput, objch <-chan *Object, resultch chan<- *SelectRecord) {
+	for object := range objch {
+		// TODO: reintroduce
+		if object.Type.IsDir() { // || errorpkg.IsCancelation(object.Err) {
+			continue
+		}
+
+		if err := object.Err; err != nil {
+			resultch <- &SelectRecord{
+				Err: object.Err,
+			}
+			continue
+		}
+
+		if object.StorageClass.IsGlacier() {
+			// TODO: this isn't fatal - how to signal?
+			resultch <- &SelectRecord{
+				Err: fmt.Errorf("object '%v' is on Glacier storage", object),
+			}
+			continue
+		}
+
+		url := object.URL
+		input.Bucket = aws.String(url.Bucket)
+		input.Key = aws.String(url.Path)
+
+		out, err := s.api.SelectObjectContentWithContext(ctx, &input)
+		if err != nil {
+			resultch <- &SelectRecord{
+				Err: err,
+			}
+		} else {
+			s.selectStream(out, resultch)
+		}
+	}
+}
+
+type Query struct {
+	ExpressionType  string
+	Expression      string
+	CompressionType string
+}
+
+func writev(to io.Writer, vector ...[]byte) (err error) {
+	for _, vec := range vector {
+		_, err = to.Write(vec)
+		if err != nil {
+			break
+		}
+	}
+	return
+}
+
+func (s *S3) Select(ctx context.Context, objch <-chan *Object, to io.Writer, query *Query, concurrency int) error {
+	if s.dryRun {
+		return nil
+	}
+
+	baseInput := s3.SelectObjectContentInput{
+		ExpressionType: aws.String(query.ExpressionType),
+		Expression:     aws.String(query.Expression),
+		InputSerialization: &s3.InputSerialization{
+			CompressionType: aws.String(query.CompressionType),
+			JSON: &s3.JSONInput{
+				Type: aws.String("Lines"),
+			},
+		},
+		OutputSerialization: &s3.OutputSerialization{
+			JSON: &s3.JSONOutput{},
+		},
+		// RequestProgress: &s3.RequestProgress{
+		// 	// TODO: conditionally enable? for larger objects? ?
+		// 	Enabled: false,
+		// },
+	}
+
+	workerCtx, cancel := context.WithCancel(ctx)
+
+	resultch := make(chan *SelectRecord, concurrency)
+	var wg sync.WaitGroup
+	wg.Add(concurrency)
+	for worker := 0; worker < concurrency; worker++ {
+		go func() {
+			defer wg.Done()
+			s.selectWorker(workerCtx, baseInput, objch, resultch)
+		}()
+	}
+
+	go func() {
+		defer close(resultch)
+		wg.Wait()
+	}()
+	// go func() {
+	// 	s.selectWorker(workerCtx, baseInput, objch, resultch)
+	// 	close(resultch)
+	// }()
+
+	var merror error
+	var fatalError error
+	for record := range resultch {
+		if record.Err == nil {
+			if fatalError != nil {
+				continue
+			}
+			if err := writev(to, record.Record, []byte{'\n'}); err != nil {
+				cancel()
+				fatalError = err
+				os.Stderr.Write([]byte(fmt.Sprintf("error writing: %v\n", err)))
+			}
+			// TODO: reinstate
+		} else if !IsCancelationError(record.Err) { //if !errorpkg.IsCancelation(record.Err) {
```

Comment:
There's a particular case where I think the network drops momentarily and a connection fails, and we don't currently retry. That causes the entire `select` command to fail.

Reply:
SDK retries all API calls. Are there other cases that we need to be aware of?

Reply:
I've seen the select command fail on terabyte-scale datasets after a few minutes due to what seemed at the time like a network error. I have a stable gigabit connection, so I'd expect this to not happen if the SDK retries API calls...

Reply:
Have you seen the error message at the time?

Reply:
I didn't record the error message, no :(